File: scripts/linters/sorted_dependencies.rb
def sorted_dependencies(root_dir, excludes = [])
  pattern = root_dir.join('crates', '*', 'Cargo.toml').to_s
  errors = Dir.glob(pattern).sort.flat_map do |path|
    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next [] if excludes.include?(relative)

    sections = parse_dep_sections(File.read(path))

    %w[dependencies dev-dependencies].filter_map do |section|
      deps = sections[section]
      next if deps.nil? || deps.empty?

      expected = sort_dep_names(deps)
      next if deps == expected

      { path: relative, section: section, actual: deps, expected: expected }
    end
  end

  return true if errors.empty?

  puts 'Found unsorted `[dependencies]` / `[dev-dependencies]` in Cargo.toml.'
  puts 'Entries must be alphabetical, with `shirabe-*` crates listed before others:'
  errors.each do |err|
    puts "  #{err[:path]} [#{err[:section]}]"
    puts "    actual:   #{err[:actual].join(', ')}"
    puts "    expected: #{err[:expected].join(', ')}"
  end
  false
end

def parse_dep_sections(content)
  sections = {}
  current = nil

  content.each_line do |line|
    stripped = line.chomp
    if stripped =~ /\A\s*\[([^\]]+)\]\s*\z/
      current = $1
      sections[current] ||= []
    elsif current && stripped =~ /\A([A-Za-z0-9_-]+)\s*[.=]/
      sections[current] << $1
    end
  end

  sections
end

def sort_dep_names(deps)
  shirabe, other = deps.partition { |d| d.start_with?('shirabe-') }
  shirabe.sort + other.sort
end

File: scripts/linters/src/Linters/SortedDependencies.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class SortedDependencies implements Linter
{
    public function name(): string
    {
        return 'sorted_dependencies';
    }

    public function failureIntro(): string
    {
        return "Found unsorted `[dependencies]` / `[dev-dependencies]` in Cargo.toml.\n"
            . 'Entries must be alphabetical, with `shirabe-*` crates listed before others:';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::cargoTomls($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            $sections = $this->parseDepSections(file_get_contents($path));

            foreach (['dependencies', 'dev-dependencies'] as $section) {
                $deps = $sections[$section] ?? [];
                if ($deps === []) {
                    continue;
                }

                $expected = $this->sortDepNames($deps);
                if ($deps === $expected) {
                    continue;
                }

                $errors[] = "{$relative} [{$section}]\n"
                    . '    actual:   ' . implode(', ', $deps) . "\n"
                    . '    expected: ' . implode(', ', $expected);
            }
        }

        return $errors;
    }

    /** @return array<string, list<string>> */
    private function parseDepSections(string $content): array
    {
        $sections = [];
        $current = null;

        foreach (explode("\n", $content) as $line) {
            $stripped = rtrim($line, "\r\n");

            if (preg_match('/\A\s*\[([^\]]+)\]\s*\z/', $stripped, $m)) {
                $current = $m[1];
                $sections[$current] ??= [];
                continue;
            }

            if ($current !== null && preg_match('/\A([A-Za-z0-9_-]+)\s*[.=]/', $stripped, $m)) {
                $sections[$current][] = $m[1];
            }
        }

        return $sections;
    }

    /**
     * @param list<string> $deps
     * @return list<string>
     */
    private function sortDepNames(array $deps): array
    {
        $shirabe = [];
        $other = [];

        foreach ($deps as $dep) {
            if (str_starts_with($dep, 'shirabe-')) {
                $shirabe[] = $dep;
            } else {
                $other[] = $dep;
            }
        }
        sort($shirabe);
        sort($other);

        return array_merge($shirabe, $other);
    }
}
